src/Mono.Android/Java.Lang/Object.cs

```diff
@@ -401,14 +401,24 @@ internal static IJavaObject PeekObject (IntPtr handle)
 					for (int i  = 0; i < wrefs.Count; ++i) {
 						var wref  = wrefs [i];
 						IJavaObject res = wref.Target as IJavaObject;
-						if (res != null && res.Handle != IntPtr.Zero && JNIEnv.IsSameObject (handle, res.Handle))
+						if (res != null && res.Handle != IntPtr.Zero && JNIEnv.IsSameObject (handle, res.Handle)) {
+							if (requiredType != null && !requiredType.IsAssignableFrom (res.GetType ())) {
+								instances.Remove (handle);
```

Comment:
It would only speed up subsequent lookups if the handle were replaced, which isn't possible to know within the context of `PeekObject()`.

Additionally, this could very easily break things in obscure ways:

```csharp
class Foo : Java.Lang.Object {}

...

var f = new Foo();
// f.Handle is registered in Object.instances
var error = JavaArray<int>.FromJniHandle (f.Handle, JniHandleOwnership.DoNotTransfer);
// f is no longer registered!
```

If `Foo` overrides any Java-side virtual methods, things will blow up the next time such a method is invoked.

Instance registration is full of peril. I can't see how this `instances.Remove()` is a good thing.